wee_slack.py

```diff
@@ -1422,11 +1423,37 @@ def update_member_presence(self, user, presence):
                 c.update_nicklist(user.id)
 
     def subscribe_users_presence(self):
-        # FIXME: There is a limitation in the API to the size of the
-        # json we can send.
-        # We should try to be smarter to fetch the users whom we want to
-        # subscribe to.
-        users = list(self.users.keys())[0:750]
+        """
+        Subscribe to the presence of users in this order (restricted by the
+        API JSON size limit):
+
+        1. Current user
+        2. Users in DMs
+        3. Users in MPDMs
+        4. Users in private groups
+        5. Users in public channels
+        """
+        dm_users, mpdm_users, group_users, channel_users = set(), set(), set(), set()
+        for channel in self.channels.values():
+            if getattr(channel, "is_open", False):
+                if channel.type == "im":
+                    dm_users.add(channel.user)
+                elif channel.type == "mpim":
+                    mpdm_users.update(channel.members)
+                elif channel.type == "group":
+                    group_users.update(channel.members)
+            elif getattr(channel, "is_member", False):
+                if channel.type == "channel":
```

Comment:
Is there a way to distinguish the mandatory channel (the one initially called #general)? Seems prudent to skip that, otherwise we just end up building a set of every user.

Reply:
Yeah, with the `is_general` attribute.

Reply:
We want to include the members from this channel as long at it's fewer than 750 though.

Reply:
If the number of members in #general are < 750, then barring any shared channels and guests, it stands to reason that the entire workspace would have <= 750 users as well, so we could just subscribe to all users without having to walk.

Reply:
Sure, but say that the entire workspace has 1000 users, but the total of all channels and dms except for the general channel is 500. Then we want to include some users from that channel.

The best might be to take members from all other channels first, and then from the general channel last. I.e. exclude the general channel from `channel_users`, but include it last in the chain.

Reply:
Perhaps. But at the same time, it seems like "presence won't be available for those in #general that you don't otherwise share channels with" is simpler to understand than "presence will be available for an arbitrary random selection of users in #general, but you can't tell who".

I dunno.

Reply:
I would probably prefer special casing smaller workspaces anyway - I've seem teams where there's a _very_ disproportionate ratio of channels to users.

Reply:
I think having a special case for <= 750 users makes sense, no need for all this work then.

Also I didn't know #general was mandatory! In that case it also makes sense to look at it last apart from the other channels.

Reply:
Re-reading the comments: I think I may actually prefer just excluding #general completely. In my workspace with > 750 users, I only share channels with 177 of them! And since #general is limited to announcements from certain users I rarely look at it and the presences of everyone there.

Reply:
Yeah, okay. You're right that having presence for just a random part of the users in general would be a bit weird, so I agree that subscribing to all users when there are less than 750 and skipping general if there are more is good.